text/000-dynamic-components.md

```diff
@@ -0,0 +1,338 @@
+
+# Summary
+
+This RFC introduces a set of principles and invariants required to preserve a reliable and predictable system while allowing lazy loading of dependencies.
+
+# Terminology
+
+**Lazy loading**: loading javascript from the server that is not initially sent to the client.
+
+**Dynamic instantiation**: creation of a new component instance where the underlying component constructor is not known until runtime.
+
+**Code splitting:** Capability to split the code base in multiple bundles, that can be loaded on demand
+
+# Basic example
+
+```html
+<template>
+    <x-lazy lwc:dynamic={customConstructor}></x-lazy>
+</template>
+```
+
+```js
+import { LightningElement, track } from "lwc";
+export default class DynamicCtor extends LightningElement {
+    @track customConstructor;
+
+    connectedCallback() {
+        this.loadCtor();
+    }
+
+    async loadCtor() {
+        const ctor = await import("c/customConstructor");
+        this.customCtor = ctor;
+    }
+}
+```
+
+# Motivation
+
+Before this RFC, LWC does not support dynamic component creation nor allow you to specify at build or design time different component configurations to pivot on during runtime. We purposely made this decision to ensure we have a clean, statically analyzable and predictable behavior on which to build upon. Due to the complexity and dynamic nature of the Lightning Platform and OSS, as we expand LWC to new clouds, applications and contexts, this strategy isn't sufficient. New primitives must be introduced to guarantee and allow for a more flexible and scalable model.
+
+# Proposal
+
+This proposal has two distinct parts (or you can think of them as two separate proposals):
+
+* The usage of the **dynamic import** to lazy load modules on the client (dynamic imports).
+* Usage of a component constructor that is unknown until runtime (template directive).
+
+
+The `import()` refers to the code splitting and the template directive `lwc:dynamic` refers to the usage of a component that is unknown at compile time in the template.
+
+## Mental model
+
+> If we could snapshot, at a given point in time, the metadata of an app and precompute all server side generated components (if any); then the whole application would be a big monolithic and static component tree, on which routes and every other pivot are simply just if/else branches.
+
+Though idealistic it is conceptually sound (and even technically achievable). It allows us to then **decide when we really need those branches** and when to trim them. Each of those branches would be lazy loaded and dynamically created.
+
+## Proposal by example
+
+Let's imagine we need to build a chart component (`lightning-chart`).
+
+A developer can choose between a bunch of types of charts (lightning-chart-histogram, lightning-chart-bar, lightning-chart-pie, ...), but we only want to expose one and branch based on a type property. We could implement it today like something like this:
```

Comment:
It may be worth being more explicit here that this scenario and following example are how things would be done pre RFC changes. I got a little confused when reading this for the first time.

btw, I love that you guys are doing these RFCs now!